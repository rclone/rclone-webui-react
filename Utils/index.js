import checkPropTypes from 'check-prop-types';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from './../src/reducers';
import {middleware} from '../src/store';

export const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
};

export const checkProps = (component, expectedProps) => {
    const propsErr = checkPropTypes(component.propTypes, expectedProps, 'prop', component.name);
    return propsErr;
};

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};