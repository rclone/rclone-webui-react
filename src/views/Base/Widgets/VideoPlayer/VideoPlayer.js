import React, {useState} from "react";
import {Button} from "reactstrap";
import * as ReactDOM from "react-dom";
import {MODAL_ROOT_ELEMENT} from "../../../../utils/Constants";
import * as PropTypes from "prop-types";
import ErrorBoundary from "../../../../ErrorHandling/ErrorBoundary";

import ReactAwesomePlayer from 'react-awesome-player'
import {connect} from "react-redux";

class PlayerComponent extends React.Component {
    // loadeddata() {
    //     console.log('loadeddata')
    // }
    // canplay() {
    //     console.log('canplay')
    // }
    // canplaythrough() {
    //     console.log('canplaythrough')
    // }
    // play() {
    //     console.log('play')
    // }
    // pause() {
    //     console.log('pause')
    // }
    // waiting() {
    //     console.log('waiting')
    // }
    // playing() {
    //     console.log('playing')
    // }
    // ended() {
    //     console.log('ended')
    // }
    error() {
        console.log('error')
    }

    render() {
        const {options} = this.props;
        return <ReactAwesomePlayer
            options={options}

            error={this.error}
        />
    }
}

function VideoPlayer({playbackURL, MimeType, loadedTestPlugins}) {

    const [preview, setPreview] = useState(true);

    const pluginUrl = loadedTestPlugins["@rclone/video-plugin"];

    // console.log("Plugin URL: " + pluginUrl, loadedTestPlugins)


    function hideFull(e) {
        e.stopPropagation();
        setPreview(!preview);

    }

    let element;
    if (preview) {
        element = (
            <div className="img-thumbnail w-100 text-center" data-test="videoPlayerWidget">
                <Button color="link" onClick={hideFull}>
                    <i className="fa fa-play-circle fa-4x"/>
                </Button>
            </div>
        )
    } else {
        if (!pluginUrl) return <p>Cannot load plugin</p>;
        // Load the video

        const subtitleURL = playbackURL.substring(0, playbackURL.lastIndexOf('.')) + ".vtt";
        console.log(subtitleURL);

        const options = {
            options: {
                sources: [{
                    type: MimeType,
                    src: playbackURL
                }],
                subtitles: [
                    {
                        language: 'en',
                        url: subtitleURL,
                        label: "EN"
                    }],
            }
        };


        element =
            ReactDOM.createPortal(
                <>
                    <div className="task-modal d-none d-sm-block" data-test="videoPlayerWidget">

                        {/*<video id="video" controls preload="metadata" width="600">*/}
                        {/*    <source src={playbackURL} type={MimeType}/>*/}
                        {/*</video>*/}
                        <PlayerComponent {...options}/>

                    </div>
                    {/*<div className="modal-backdrop fade show"></div>*/}


                </>, document.getElementById(MODAL_ROOT_ELEMENT));
    }

    return (
        <ErrorBoundary>
            {element}
        </ErrorBoundary>
    );


}

VideoPlayer.propTypes = {
    playbackURL: PropTypes.string.isRequired,
    MimeType: PropTypes.string.isRequired,
    loadedTestPlugins: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
    loadedTestPlugins: state.plugins.loadedTestPlugins
});

export default connect(mapStateToProps, {})(VideoPlayer);
