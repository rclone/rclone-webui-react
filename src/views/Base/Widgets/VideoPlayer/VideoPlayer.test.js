import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../../../Utils";
import toJson from "enzyme-to-json";
import VideoPlayer from "./VideoPlayer";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);
    const component = shallow(<VideoPlayer {...props} store={store}/>);
    return component.childAt(0).dive();
};


describe('Video Player', function () {


    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {};


            const props = {
                playbackURL: '',
                MimeType: ''
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "videoPlayerWidget");
            expect(component).toHaveLength(1);
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });

    });

});