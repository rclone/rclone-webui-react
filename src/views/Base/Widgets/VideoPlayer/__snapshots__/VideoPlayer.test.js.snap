// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Player renders should match snapshot 1`] = `
<ErrorBoundary>
  <div
    className="img-thumbnail w-100 text-center"
    data-test="videoPlayerWidget"
  >
    <Button
      color="link"
      onClick={[Function]}
      tag="button"
    >
      <i
        className="fa fa-play-circle fa-4x"
      />
    </Button>
  </div>
</ErrorBoundary>
`;
